### PR TITLE
[SPARK-45652][SQL][3.4] SPJ: Handle empty input partitions after dynamic filtering

### PR DESCRIPTION
**Pull request by sunchao**
This is a cherry-pick of https://github.com/apache/spark/pull/43531 to branch-3.4, with a few modifications.

### What changes were proposed in this pull request?

Handle the case when input partitions become empty after V2 dynamic filtering, when SPJ is enabled.

### Why are the changes needed?

Current in the situation when all input partitions are filtered out via dynamic filtering, SPJ doesn't work but instead will panic:
```
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:529)
	at scala.None$.get(Option.scala:527)
	at org.apache.spark.sql.execution.datasources.v2.BatchScanExec.filteredPartitions$lzycompute(BatchScanExec.scala:108)
	at org.apache.spark.sql.execution.datasources.v2.BatchScanExec.filteredPartitions(BatchScanExec.scala:65)
	at org.apache.spark.sql.execution.datasources.v2.BatchScanExec.inputRDD$lzycompute(BatchScanExec.scala:136)
	at org.apache.spark.sql.execution.datasources.v2.BatchScanExec.inputRDD(BatchScanExec.scala:135)
```

This is because the `groupPartitions` method will return `None` in this scenario. We should handle the case.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a test case for this.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by sunchao** [comment]
cc @HyukjinKwon @viirya @dongjoon-hyun @cloud-fan

**Review by dongjoon-hyun — COMMENTED** [review]
Do we have this at `branch-3.5`, @sunchao ? We cannot merge this if we don't have this in `branch-3.5` because it will cause a regression when the customers upgrade from 3.4 to 3.5.

**Comment by sunchao** [comment]
@dongjoon-hyun oops I forgot branch-3.5. Just opened https://github.com/apache/spark/pull/43540.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. (Pending CIs.) Thank you, @sunchao .

**Review by viirya — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @sunchao and @viirya . Merged to master.

**Pull request closed without merging**
